Source/Core/Common/MemArenaWin.cpp

```diff
@@ -16,60 +17,395 @@
 #include "Common/MsgHandler.h"
 #include "Common/StringUtil.h"
 
+using PVirtualAlloc2 = PVOID(WINAPI*)(HANDLE Process, PVOID BaseAddress, SIZE_T Size,
+                                      ULONG AllocationType, ULONG PageProtection,
+                                      MEM_EXTENDED_PARAMETER* ExtendedParameters,
+                                      ULONG ParameterCount);
+
+using PMapViewOfFile3 = PVOID(WINAPI*)(HANDLE FileMapping, HANDLE Process, PVOID BaseAddress,
+                                       ULONG64 Offset, SIZE_T ViewSize, ULONG AllocationType,
+                                       ULONG PageProtection,
+                                       MEM_EXTENDED_PARAMETER* ExtendedParameters,
+                                       ULONG ParameterCount);
+
+using PIsApiSetImplemented = BOOL(APIENTRY*)(PCSTR Contract);
+
 namespace Common
 {
+struct WindowsMemoryRegion
+{
+  u8* m_start;
+  size_t m_size;
+  bool m_is_mapped;
+
+  WindowsMemoryRegion(u8* start, size_t size, bool is_mapped)
+      : m_start(start), m_size(size), m_is_mapped(is_mapped)
+  {
+  }
+};
+
 MemArena::MemArena()
 {
+  // Check if VirtualAlloc2 and MapViewOfFile3 are available, which provide functionality to reserve
+  // a memory region no other allocation may occupy while still allowing us to allocate and map
+  // stuff within it. If they're not available we'll instead fall back to the 'legacy' logic and
+  // just hope that nothing allocates in our address range.
+  HMODULE kernelBaseHandle = LoadLibraryW(L"KernelBase.dll");
```

Comment:
Can we use DynamicLibrary (from Common) here instead?

Reply:
Oh, I didn't realize we had that! Yes, probably, let's see...